### DISCUSSION TITLE
Multiple source files with `link_program`

### DISCUSSION THREAD

**Discussion by Richardk2n** [pyopencl/pyopencl#598] [Q&A] [answered]
If I understood correctly `link_program` is the linker step that allows me to combine the compiled binaries of multiple source files together.

I managed to get the source to compile with `compile` (also with headers).
For a single source example `link_program` also works for me.

As soon as I try to use multiple programs tho, `link_program` does give me an error (LINK_PROGRAM_FAILURE).

The documentation does not really cover this in detail and I am not sure what I have to compile, what I should not compile, how the source files may look (e.g. limitations that may apply) and what to pass to `link_program`. Also how does `-create_library` factor into this?

If you could give me a minimal working example with multiple source files, that would be amazing.

**Comment by inducer** [comment, marked as answer]
How about this?

https://github.com/inducer/pyopencl/blob/3a96299f0e60248aa11261646d82155c0c6bbd89/test/test_wrapper.py#L1161-L1195
